### ISSUE TITLE
member.roles.add() doesn't always work

### ISSUE THREAD

**Issue by ItIsN3RD** [discordjs/discord.js#4930]
**Please describe the problem you are having in as much detail as possible:**
Sometimes it works and adds the role other times it just doesnt do anything, it does not throw any errors.

**Include a reproducible code sample here, if possible:**

```js
// MY CODE 
bot.on('guildMemberAdd', member => {
        bot.channels.cache.get('766305878172762203').send('<@'+member.id+'>');
        
        let role = member.guild.roles.cache.get("741484763013185586");

        if (member.roles.cache.find(r => r.name !== "Member")) {
                member.roles.add(role);
                console.log(member.id);
        }
});
```

**Further details:**

- discord.js version: 12.4.0 current and working better than 12.3.1.
- Node.js version: npm@6.14.8
- Operating system: Windows 10
- Priority this issue should have – please be realistic and elaborate if possible: To be honest I think adding roles is a pretty big issue and should be prioritized over most things. 

**Relevant client options:**

- partials: none
- gateway intents: none
- other: none

- [ ] I have also tested the issue on latest master, commit hash:

**Comment by NotSugden** [comment]
this issue tracker is for actual bugs or feature requests, not issues with your code, (try the [discord server](https://discord.gg/bRCvFy9)), you should also explain what your problem is "doesn't work" says nothing about your problem, is your code unemployed? does it not add the role? etc

**Comment by ItIsN3RD** [comment]
@almostSouji I'm 99% sure its not my code. Sorry but I dont know what else to say other than it "doesnt always work", literally it works under 25% of the time. The on guildmemberadd event catches everyones IDs that join its just the discord API doesn't always add the person to a role.

**Comment by almostSouji** [comment]
Alright, I'm going to re-open and see if someone else can reproduce this, because I can not.
Without specifics as to under which circumstances this happens a reproduction is unlikely, will see if there are any silent returns possible (there really shouldn't be)

If the role is not added what happens? Any errors? Or just nothing? Does the Promise still resolve?

**Comment by udit2303** [comment]
> @almostSouji @NotSugden I'm 99% sure its not my code. Sorry but I dont know what else to say other than it "doesnt always work", literally it works under 25% of the time. The on guildmemberadd event catches everyones IDs that join its just the discord API doesn't always add the person to a role.

Try downgrading to v12.3.1 by `npm install discord.js@12.3.1` and see if you can reproduce this. I haven't been having this issue with my code. I also made another bot to check this and it seems to work fine. Also what did you you want to do with `if (member.roles.cache.find(r => r.name !== "Member"))` condition. Are you trying to check if the name of the role is not member or something that I didn't understand?

**Comment by almostSouji** [comment]
While, yes, that's wrong - that'll also always be true, after all everyone always has the `@everyone` role, meaning it's not an issue for the execution of this scope.

Downgrading is not really going to help with trying to find out under which circumstances it "doesn't always work" in 12.4

**Comment by ItIsN3RD** [comment]
@udit2303 Basically what I was trying to do is add the person and then do an if statement to check to see if they had the role member yet and if they didn't I would try to add them again. However I did try version 12.3.1 and was still acting up so I decided to update to the latest discord.js version and upload my code to Heroku and it seems to be working fine ill keep you guys posted in a few days to see if that solved the issue, I wonder if having 20+ discord bots running on my pc was starting to slow stuff down and making stuff not work properly.   
```
member.roles.add(role);
// Then do this if statement to check to see if it didnt work and try to re add them hopfully
if (member.roles.cache.find(r => r.name !== "Member")) {
                member.roles.add(role);
                console.log(member.id);
        }
```

**Comment by wasdennnoch** [comment]
Adding the role is an async operation which you have to await. If you just call the function and then immediately after do a check if the member has the role yet, that check will return false as the async network request hasn't succeeded yet.
As this seems to be an issue with understanding/working with Javascript Promises this is a question for the [discord server](https://discord.gg/bRCvFy9), not the issue tracker.

**Comment by tipakA** [comment]
That is... completely not how that works. That code is run every time the `guildMemberAdd` event is emitted - so once per member, assuming the member doesn't leave and rejoin (upon which they would not have your role anyway).
It won't be emitted again for unknown reasons, so there is no point in checking if they don't have the member role, because they do not have it - they just joined.

Additionally, that is completely not how you check if someone does not have a specific role. What your `if` does, is it searches for a role which name is not equal to `Member` - `member`, `admin` or, as already mentioned `@everyone` roles will be found by that if. And considering that *every* member has the `@everyone` role at all timess, your if will *always* run.

There still isn't really much explanation or context to your issue - so far everything we know is that "it sometimes doesn't work", and that it apparently still occurs on 12.3.1.
Do you have other bots that might interfere? Is it always same person, and if someone leaves and joins are they still not getting the role? How big is the bot?

<hr>

@wasdennnoch Seems that you didn't really read what the code actually does despite what the person who wrote it claims it does.

**Comment by ItIsN3RD** [comment]
I literally don't know what's going on I didn't change any of my code now its working flawlessly.

**Comment by snnaakee** [comment]
This issue is very problematic and still happening to me since months.
Tried everything to fix it, still nothing, the role is added, then the next time it doesn't work, then it work again... 
When you google the issue, it seems to happen to a lot of people.

`client.on('guildMemberAdd', member => {
var i = "336796839582105600";
var role = member.guild.roles.cache.find(r => r.id === i);
member.roles.add(role).then(console.log('Role added: ' + member.id));
`

**Comment by PanSzelescik** [comment]
I have a little hack, fetch member with `guild.members.fetch({user: id, force: true})` and then try add/remove roles. With this i don't have any problems

**Comment by almostSouji** [comment]
> The audits shows the bot adding the role, but when you look at the new user's profile, no role shows up.

That suggests this is a client issue. I have had that issue myself, where the client doesn't show bot-applied roles. However, API wise the request goes through, is fulfilled and the role accordingly added. There is nothing you (or we, on the library level) can do to influence the viewing client.

This is accordingly not a library issue, and the above mentioned "little hack" completely unrelated to the question or topic at hand.

**Comment by Floliroy** [comment]
Hi, i'm having the same issue and i cant find any solution, any help please ?

**Comment by tarolling** [comment]
Also having the same issue, I have mine inside a `.then()` statement but even then it works about 75% of the time which isn't consistent enough.

**Comment by Jstanford5216** [comment]
Likewise I have this issue as well just to raise awareness a bit more and mine is completely awaited and the request goes through but the user simply doesn't get the roles added sometimes and other times it works fine and it isn't ratelimit or anything like that as I don't use it that often.

**Comment by almostSouji** [comment]
If you can still reproduce this on `discord.js@dev` please open a separate issue and provide the exact version (`discord.js@13.0.0-dev.commithash.timestamp`, you get this from `npm list discord.js`) and detailed steps on how to reproduce this issue, including the code you run in the format of a minimal, reproducible code sample.

If you think it's related to this issue feel free to mention that in the issue body.

---

Please provide a minimal reproducible code sample so we can attempt to reproduce and verify your issue. This means isolating the erroneous behavior to a fresh bot, or stripping away functionality until you can be sure the code is minimal. Also, include a manual with the steps you take interacting with the code in question which results in the behavior you outlined.

Please try to reproduce the issue yourself on a fresh bot. If you can't reproduce it, it's very likely that we can't either and the code sample in question is not enough.

Boilerplate: https://discord.js.org/#/docs/main/stable/examples/ping
Minimal Reproducible sample: https://stackoverflow.com/help/minimal-reproducible-example

**Comment by snnaakee** [comment]
You failed to reproduce it because the issue appear to be present when x members have been reached in that discord server. I assume x to be 1000, but it could be less. 
The event is fired, but the role is never granted to the client. You can loop it and add the role thousands times, no role will be added.
My post on [stackoverflow](https://stackoverflow.com/questions/66457086/discord-js-adding-role-upon-guildmemberadd-doesnt-always-work) while ago did not brought any answer either.

The creator of this issue already provided a very good reproducible example, all you need to add is 1000 members to your discord server to reproduce it.

**Comment by monbrey** [comment]
> You can loop it and add the role thousands times

This seems like a very good way to reproduce the error - it's called a *rate limit* and it will prevent the API from responding, if not get your bot banned.

**Comment by snnaakee** [comment]
> > You can loop it and add the role thousands times
> 
> This seems like a very good way to reproduce the error - it's called a _rate limit_ and it will prevent the API from responding, if not get your bot banned.

Please just stay focus on the issue we are trying to bring. It's already hard to get attention on it and we are dealing with this crap  for years now. Adding roles manually because the discord API failed to grant them properly.
I obviously over reacted when mentioning "thousands times", you get the idea.

**Comment by monbrey** [comment]
Souji's comment was focused on this issue - he asked if you could reproduce it on the latest discord.js release.

If your focus is still on the role appearing in the client rather than the execution of API calls, then the issue doesn't appear to lie with discord.js, and you should follow this up with Discord directly.

Otherwise, I've just tried to reproduce this myself also, and we aren't able to with any of the code samples provided in this thread. If you still can on the latest `discord.js@dev` release, please provide a sample that does.

**Comment by DTrombett** [comment]
Well, what here is missing is a working example with a good circumstance. If you are sure that you're able to reproduce this bug (which actually doesn't seem to be related with djs itself) try adding a `console.log(member.user.tag)` at the start of the event, and if you see that a member doesn't receive the role, check that the console logged the user tag. If it wasn't, so most probably Discord didn't send the event.

**Comment by ScreamingHawk** [comment]
Stopping by to say I'm also having this issue with `discord.js v13.6.0`

**Comment by snnaakee** [comment]
That won't help, but I've been using Discord.NET ever since and haven't had the problem again.

**Comment by ImRodry** [comment]
> Stopping by to say I'm also having this issue with `discord.js v13.6.0`. And for bulk removing roles too

please make a new issue with a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) as this was closed over a year ago

**Comment by ScreamingHawk** [comment]
Fixed it! 
Going to reply here because this is the first result that appears on Google. Gotta love SEO... 

Couple steps:
1. Make sure you are using the most up to date user information
2. Only remove roles the user does already have
3. Only add roles the user doesn't already have

```js
		member = await guild.members.fetch({ user: discordId, force: true })

		const toRemove = removeList.filter(
			r => member.roles.cache.has(r),
		)
		if (toRemove.length) {
			await member.roles.remove(toRemove, `Removing roles`)
		}

		const toAdd = addList.filter(
			r => !member.roles.cache.has(r),
		)
		if (toAdd .length) {
			await member.roles.add(toAdd , `Adding roles`)
		}
```

**Comment by almostSouji** [comment]
Fix it even better and do it all at once via `GuildMember#edit`. 
The issue you describe is mostly present due to race conditions with multiple patch calls on a single guild member (the endpoint called by all these methods). This is even visible when multiple users modify the roles of a single user, which becomes much more pronounced with bots though.

TL;DR: `GuildMember#edit` once, don't `#add` / `#remove` per role you want to remove.
